File: Wire-iOS/Sources/UserInterface/UserProfile/Views/ParticipantDeviceHeaderView.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "ParticipantDeviceHeaderView.h"
#import "ParticipantDeviceHeaderView+Internal.h"
@import PureLayout;
#import "WebLinkTextView.h"
#import "WireExtensionComponents.h"
#import "NSAttributedString+Wire.h"
#import "Wire-Swift.h"

@interface ParticipantDeviceHeaderView () <UITextViewDelegate>
@property (strong, nonatomic) UITextView *textView;
@property (strong, nonatomic, readwrite) NSString *userName;
@end



@implementation ParticipantDeviceHeaderView

- (instancetype)initWithUserName:(NSString *)userName
{
    self = [super initWithFrame:CGRectZero];
    
    if (self) {
        _userName = userName;
        [self setup];
    }
    return self;
}

- (void)setup
{
    self.backgroundColor = UIColor.clearColor;
    [self createViews];
    [self setupConstraints];

    [self setupStyle];
}

- (void)createViews
{
    self.textView = [[WebLinkTextView alloc] init];
    self.textView.textContainer.maximumNumberOfLines = 0;
    self.textView.delegate = self;
    self.textView.linkTextAttributes = @{};
    
    [self addSubview:self.textView];
}

- (void)setShowUnencryptedLabel:(BOOL)showUnencryptedLabel
{
    self.textView.attributedText = [self attributedExplanationTextForUserName:self.userName showUnencryptedLabel:showUnencryptedLabel];
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    [self.textView setNeedsUpdateConstraints];
    [self.textView updateConstraintsIfNeeded];
}

- (void)setupConstraints
{
    [self.textView autoPinEdgesToSuperviewEdgesWithInsets:UIEdgeInsetsMake(40, 24, 16, 24)];
}

#pragma mark - UITextViewDelegate

- (BOOL)textView:(UITextView *)textView shouldInteractWithURL:(NSURL *)URL inRange:(NSRange)characterRange interaction:(UITextItemInteraction)interaction
{
    if ([self.delegate respondsToSelector:@selector(participantsDeviceHeaderViewDidTapLearnMore:)]) {
        [self.delegate participantsDeviceHeaderViewDidTapLearnMore:self];
    }
    return NO;
}

#pragma mark - Attributed Text

- (NSAttributedString *)attributedExplanationTextForUserName:(NSString *)userName showUnencryptedLabel:(BOOL)unencrypted
{
    if (unencrypted) {
        NSString *message = NSLocalizedString(@"profile.devices.fingerprint_message_unencrypted", nil);
        return [self attributedFingerprintForUserName:userName message:message];
    }
    else {
        NSString *message = [NSString stringWithFormat:@"%@%@", NSLocalizedString(@"profile.devices.fingerprint_message.title", nil), NSLocalizedString(@"general.space_between_words", nil)];

        NSMutableAttributedString * mutableAttributedString = [[NSMutableAttributedString alloc] initWithAttributedString: [self attributedFingerprintForUserName:userName message:message]];

        NSString *fingerprintLearnMoreLink = NSLocalizedString(@"profile.devices.fingerprint_message.link", nil);

        [mutableAttributedString appendString:fingerprintLearnMoreLink attributes:[self linkAttributes]];

        return mutableAttributedString;
    }
}

- (NSMutableParagraphStyle *)paragraphStyleForFingerprint
{
    NSMutableParagraphStyle *paragraphStyle = NSParagraphStyle.defaultParagraphStyle.mutableCopy;
    paragraphStyle.lineSpacing = 2;

    return paragraphStyle;
}

- (NSAttributedString *)attributedFingerprintForUserName:(NSString *)userName message:(NSString *)message
{
    NSString *fingerprintExplanation = [NSString stringWithFormat:message, userName];

    NSMutableParagraphStyle *paragraphStyle = [self paragraphStyleForFingerprint];

    NSDictionary *textAttributes = @{
                                     NSForegroundColorAttributeName: self.textColor,
                                     NSFontAttributeName: self.font,
                                     NSParagraphStyleAttributeName: paragraphStyle
                                     };

    NSMutableAttributedString *attributedText = [[NSMutableAttributedString alloc] initWithString:fingerprintExplanation
                                                                                       attributes:textAttributes];

    return [[NSAttributedString alloc] initWithAttributedString:attributedText];
}

- (NSDictionary *)linkAttributes
{
    NSMutableParagraphStyle *paragraphStyle = [self paragraphStyleForFingerprint];

    NSDictionary *linkAttributes = @{
                                     NSFontAttributeName: self.font,
                                     NSForegroundColorAttributeName: self.linkAttributeColor,
                                     NSLinkAttributeName: NSURL.wr_fingerprintLearnMoreURL,
                                     NSParagraphStyleAttributeName: paragraphStyle
                                     };

    return linkAttributes;
}

@end

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Views/ParticipantDeviceHeaderView.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit
import WireDesign

// MARK: - ParticipantDeviceHeaderViewDelegate

protocol ParticipantDeviceHeaderViewDelegate: AnyObject {
    func participantsDeviceHeaderViewDidTapLearnMore(_ headerView: ParticipantDeviceHeaderView)
}

// MARK: - ParticipantDeviceHeaderView

final class ParticipantDeviceHeaderView: UIView {
    // MARK: Lifecycle

    init(userName: String) {
        self.userName = userName

        super.init(frame: CGRect.zero)

        setup()
    }

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: Internal

    let userName: String

    weak var delegate: ParticipantDeviceHeaderViewDelegate?

    var showUnencryptedLabel = false {
        didSet {
            textView.attributedText = attributedExplanationText(
                for: userName,
                showUnencryptedLabel: showUnencryptedLabel
            )
        }
    }

    var linkAttributes: [NSAttributedString.Key: Any] {
        [
            NSAttributedString.Key.font: font,
            NSAttributedString.Key.foregroundColor: linkAttributeColor,
            NSAttributedString.Key.link: WireURLs.shared.whyToVerifyFingerprintArticle,
            NSAttributedString.Key.paragraphStyle: paragraphStyleForFingerprint,
        ]
    }

    var paragraphStyleForFingerprint: NSMutableParagraphStyle {
        let paragraphStyle = NSParagraphStyle.default.mutableCopy() as! NSMutableParagraphStyle
        paragraphStyle.lineSpacing = 2

        return paragraphStyle
    }

    func setup() {
        backgroundColor = .clear
        setupViews()
        setupConstraints()
        setupAccessibility()
    }

    func setupViews() {
        textView.textContainer.maximumNumberOfLines = 0
        textView.delegate = self
        textView.linkTextAttributes = [:]

        addSubview(textView)
    }

    // MARK: Attributed Text

    func attributedExplanationText(
        for userName: String,
        showUnencryptedLabel unencrypted: Bool
    ) -> NSAttributedString? {
        typealias ProfileDevices = L10n.Localizable.Profile.Devices

        if unencrypted {
            let message = userName.isEmpty ? ProfileDevices.noDeviceData : ProfileDevices
                .fingerprintMessageUnencrypted(userName)
            return attributedFingerprint(forMessage: message)
        } else {
            let message =
                "\(ProfileDevices.FingerprintMessage.title(userName))\(L10n.Localizable.General.spaceBetweenWords)"

            let mutableAttributedString =
                NSMutableAttributedString(attributedString: attributedFingerprint(forMessage: message))

            let fingerprintLearnMoreLink = ProfileDevices.FingerprintMessage.link && linkAttributes

            return mutableAttributedString + fingerprintLearnMoreLink
        }
    }

    func attributedFingerprint(forMessage fingerprintExplanation: String) -> NSAttributedString {
        let textAttributes = [
            NSAttributedString.Key.foregroundColor: textColor,
            NSAttributedString.Key.font: font,
            NSAttributedString.Key.paragraphStyle: paragraphStyleForFingerprint,
        ]

        return NSAttributedString(string: fingerprintExplanation, attributes: textAttributes)
    }

    // MARK: Private

    private var font: UIFont = .normalLightFont
    private var textColor: UIColor = SemanticColors.Label.textSectionHeader
    private var linkAttributeColor: UIColor = .accent()
    private var textView = WebLinkTextView()

    private func setupConstraints() {
        textView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            textView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 24),
            textView.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -24),
            textView.topAnchor.constraint(equalTo: topAnchor, constant: 16),
            textView.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -16),
        ])
    }

    private func setupAccessibility() {
        textView.accessibilityTraits = .link
        textView.accessibilityHint = L10n.Accessibility.DeviceDetails.WhyVerifyFingerprint.hint
    }
}

// MARK: UITextViewDelegate

extension ParticipantDeviceHeaderView: UITextViewDelegate {
    func textView(
        _ textView: UITextView,
        shouldInteractWith URL: URL,
        in characterRange: NSRange,
        interaction: UITextItemInteraction
    ) -> Bool {
        delegate?.participantsDeviceHeaderViewDidTapLearnMore(self)

        return false
    }
}
